vault/resource_database_secret_backend_static_role.go

```diff
@@ -0,0 +1,210 @@
+package vault
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/vault/api"
+)
+
+var (
+	databaseSecretBackendStaticRoleBackendFromPathRegex = regexp.MustCompile("^(.+)/static-roles/.+$")
+	databaseSecretBackendStaticRoleNameFromPathRegex    = regexp.MustCompile("^.+/static-roles/(.+$)")
+)
+
+func databaseSecretBackendStaticRoleResource() *schema.Resource {
+	return &schema.Resource{
+		Create: databaseSecretBackendStaticRoleWrite,
+		Read:   databaseSecretBackendStaticRoleRead,
+		Update: databaseSecretBackendStaticRoleWrite,
+		Delete: databaseSecretBackendStaticRoleDelete,
+		Exists: databaseSecretBackendStaticRoleExists,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Unique name for the static role.",
+			},
+			"backend": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The path of the Database Secret Backend the role belongs to.",
+			},
+			"username": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The database username that this role corresponds to.",
+			},
+			"rotation_period": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: "The amount of time Vault should wait before rotating the password, in seconds.",
+				ValidateFunc: func(v interface{}, k string) (ws []string, errs []error) {
+					value := v.(int)
+					if value < 5 {
+						errs = append(errs, fmt.Errorf("The minimum value of rotation_period is 5 seconds."))
+					}
+					return
+				},
+			},
+			"db_name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Database connection to use for this role.",
+			},
+			"rotation_statements": {
+				Type:        schema.TypeList,
+				Required:    true,
```

Comment:
I noticed on the docs that this isn't required on Vault's end. Was there a reason for requiring it here?

Reply:
I thought so too initially, however, while testing with bellow commands I discovered it actually is required.

Started postgres  container in tab 1.
```sh
docker run -it --rm \
    -p 5432:5432 \
    -e POSTGRES_PASSWORD=secret \
    -e POSTGRES_USER=root \
    postgres
```

Started vault in tab 2.
```sh
vault server -dev
```

Run following commands in tab 3.
```sh
vault secrets enable database

vault write database/config/my-postgresql-database \
    plugin_name=postgresql-database-plugin \
    allowed_roles="my-role" \
    connection_url="postgresql://{{username}}:{{password}}@127.0.0.1:5432/?sslmode=disable" \
    username="root" \
    password="secret"

vault write database/static-roles/my-role \
    username="test" \
    rotation_period="24h" \
    db_name="my-postgresql-database"
```

When executing last command Vault returns error:
```
Error writing data to database/static-roles/my-role: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/database/static-roles/my-role
Code: 500. Errors:

* 1 error occurred:
	* error setting credentials: empty rotation statements
```

I guess we could pass the `rotation_statements` parameter to backend, but leave it empty by default. Should we do that?

Reply:
Ah, I see! Interesting. When I saw that output, I thought it actually might be a bug on Vault's end, so I poked at it a little. I found that for the postgres engine, the rotation statements are required [here](https://github.com/hashicorp/vault/blob/master/plugins/database/postgresql/postgresql.go#L103). _But_, for [the MSSQL secrets engine](https://github.com/hashicorp/vault/blob/master/plugins/database/mssql/mssql.go), "statements.Rotation" aren't used at all, and in [the MySQL secrets engine](https://github.com/hashicorp/vault/blob/master/plugins/database/mysql/mysql.go#L275), they're defaulted if they're not provided. So, the behavior is particular to each back-end, which puts us in an unusual spot.

That makes me think that maybe we should leave it marked as optional, only include it in calls if it's set, pass it through to Vault, and immediately return any error that Vault gives. Does that seem like the best way to handle it to you? Or are there potential problems I'm not thinking of?

Reply:
Sure, so the code remains as it currently is, but I need to fix the tests, because acceptance tests use Postgres we are required to set the rotation statements. We could change acceptance to use MySQL, or even add additional cases to test on both Postgres and MySQL, WDYT?

Reply:
Hmmm, for a long time, I've been wanting to add the container for at least one database to the tests. I think MySQL would be a good one to start with. Let me see if I can do it just now, and if so, I'll PR it to master so it can be used here after it's merged.

If that holds you up, though, feel free to take a different approach, though one doesn't occur to me off-hand.